### DISCUSSION TITLE
How to connect PG via ipv6

### DISCUSSION THREAD

**Discussion by Jason7602** [Kong/kong#7304] [Ideas and feature requests]
### Summary
connect to pg via ipv6 encountered such problem
SUMMARY_GOES_HERE

### Steps To Reproduce

1.
2.
3.
4.

### Additional Details & Logs

- Kong version (`$ kong version`)
1.5.1
- Kong debug-level startup logs (`$ kong start --vv`)
- Kong error logs (`<KONG_PREFIX>/logs/error.log`)
Error: [PostgreSQL error] failed to retrieve PostgreSQL server_version_num: No address associated with hostname

- Kong configuration (the output of a GET request to Kong's Admin port - see
  https://docs.konghq.com/latest/admin-api/#retrieve-node-information)
- Operating system

**Comment by florian-ma** [comment]
Have found the same issue on IPv6 only system.

- Kong version 2.1.2
- Using a database accessible by Postgres tools from Kong's location
- Tried with and without squared brackets for the KONG_PG_HOST value
- Same error message "[PostgreSQL error] failed to retrieve PostgreSQL server_version_num: No address associated with hostname"

After looking into Kong directly for versions 2.1.2 (used) and 2.2.1 (latest), it looks like the pgmoon library is just not compatible with IPv6. It is never mentioned at all in the latest version 1.11.0, also used by Kong.
https://github.com/leafo/pgmoon

**Reply by jzdwk** [reply]
pgmoon it self has three types of socket. see 
[pgmoon](https://github.com/leafo/pgmoon#dependencies) , and if using socket_type = "nginx" in openresty, it does work. 

```
   location /pgtest {
        default_type text/html;
        content_by_lua_block {
                local pgmoon = require("pgmoon")
                local pg = pgmoon.new({
                  host = "[8888::a99]",
                  port = "5432",
                  database = "test",
                  user = "test",
                  password = "****",
                  socket_type = "nginx"
                })
                ngx.log(ngx.INFO, "finish new pg config")
                success, err = pg:connect()
                if err then
                  ngx.log(ngx.ERR, "connect pg err:", err)
                  ngx.say(err)
                else
                  ngx.say("connect success")
                end
      }
   }
```
so is it possible to change socket_type in kong? [connect](https://github.com/Kong/kong/blob/ec7ac810e7d6a5f6635edec6f05b94ec6c424589/kong/db/strategies/postgres/connector.lua#L195)

**Reply by mayocream** [reply]
I think pgmoon's `socket_type=nginx` mode uses OpenResty's functions which support ipv6.
But during `init` phase or when running CLI, Kong is using  "pure" Lua, in this case, it has to be `luasocket` mode.

**Reply by jzdwk** [reply]
thanks for your reply, and i find that luasocket is compatible with ipv6 by using socket.tcp6(), but pgmoon only  uses tcp4, see [pgmoon socket.lua](https://github.com/leafo/pgmoon/blob/601c1cb6cf37579b854c89b55423899f03962b5a/pgmoon/socket.lua#L30). so i try to make changes in pgmoon and rebuild kong:
```
  -- at socket.lua 
  luasocket = {
    tcp = function(...)
      local socket = require("socket")
      --add tcp6 support
      local sock = socket.tcp6(...)
      ...
```
and kong start successfully:)
```
docker run -d --name kong-host \
     ....
     -e "KONG_DATABASE=postgres" \
     -e "KONG_PG_HOST=8888::a99" \
     ....
     kong-v6:luasocket

[root@VM-24-5-centos hostmode]# docker logs kong-host --until 50m
2022/01/14 08:53:54 [debug] 1#0: [lua] globalpatches.lua:10: installing the globalpatches
2022/01/14 08:53:54 [debug] 1#0: [lua] globalpatches.lua:246: randomseed(): seeding PRNG from OpenSSL RAND_bytes()
2022/01/14 08:53:54 [debug] 1#0: [lua] globalpatches.lua:272: randomseed(): random seed: 189221072262 for worker nb 0
...
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: api-transformer
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: correlation-id
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: pre-function
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: cors
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: ldap-auth
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: loggly
2022/01/14 08:53:55 [debug] 1#0: [lua] plugins.lua:247: load_plugin(): Loading plugin: hmac-auth

```

**Reply by mayocream** [reply]
Great! I think you can create a PR to pgmoon!
